game-core/src/main/java/games/strategy/triplea/delegate/battle/TargetGroup.java

```diff
@@ -0,0 +1,90 @@
+package games.strategy.triplea.delegate.battle;
+
+import com.google.common.collect.Sets;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.Matches;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.AccessLevel;
+import lombok.Getter;
+import org.triplea.java.collections.CollectionUtils;
+
+/** Group of firing units and their targets. */
+@Getter(AccessLevel.PRIVATE)
+public class TargetGroup {
+
+  private final Set<UnitType> firingUnitTypes;
+  private final Set<UnitType> targetUnitTypes;
+
+  public TargetGroup(final UnitType firingUnitType, final Set<UnitType> targetUnitTypes) {
+    firingUnitTypes = Sets.newHashSet(firingUnitType);
+    this.targetUnitTypes = targetUnitTypes;
+  }
+
+  public Collection<Unit> getFiringUnits(final Collection<Unit> units) {
+    return CollectionUtils.getMatches(units, Matches.unitIsOfTypes(firingUnitTypes));
+  }
+
+  public Collection<Unit> getTargetUnits(final Collection<Unit> units) {
+    return CollectionUtils.getMatches(units, Matches.unitIsOfTypes(targetUnitTypes));
+  }
+
+  /**
+   * Find all target groupings for firing units and enemy units based on canNotTarget and
+   * canNotBeTargetedBy attributes. Also takes into account if any destroyers are present which
+   * cancel canNotBeTargetedBy. Sort all the target groups so those with the least number of targets
+   * appear first.
+   */
+  public static List<TargetGroup> newTargetGroups(
+      final Collection<Unit> units, final Collection<Unit> enemyUnits) {
```

Comment:
I see AI is tearing through this method quite frequently while simulating battles. Would it make sense to early return an empty list here if either of the input parameters are empty? I'm wondering if that might help performance a bit, particularly if unit stacks get to be very large, and secondly it might help a code maintainer grok this method to quickly know this method is effectively a no-op if either parameter is empty.

Reply:
Uh we could though I don't expect it to have much performance impact (probably <1%). If you did some tests and saw differently then I'd definitely look to add that in.

Reply:
Yeah, I don't think there would be a huge performance win, maybe if the stack sizes were very large. It's more an early termination to make it easier for a reader to see when the method is a no-op, similar to the if-guard you have in another method (line 74)
```
    if (targets.isEmpty()) {
      return;
    }
```